src/gcp/cloudsql/permissions.ts

```diff
@@ -221,22 +276,316 @@ export async function setupSQLPermissions(
+    (owner) => owner !== firebaseOwnerRole,
+  );
+  const grantCmds = nonFirebasetablesOwners.map(
+    (owner) => `GRANT "${firebasewriter(databaseId, schema)}" TO "${owner}"`,
+  );
+
+  // Step 3: Alter table owners permissions
+  const alterTableCmds = schemaInfo.tables.map(
+    (table) => `ALTER TABLE "${schema}"."${table.name}" OWNER TO "${firebaseOwnerRole}";`,
+  );
+
+  // Run sql commands
+  await executeSqlCmdsAsSuperUser(
+    options,
+    instanceId,
+    databaseId,
+    [...grantCmds, ...alterTableCmds],
+    silent,
+  );
+}
+
+export async function brownfieldSqlSetup(
+  instanceId: string,
+  databaseId: string,
+  schemaInfo: SchemaMetaData,
+  options: Options,
+  silent: boolean = false,
+) {
+  const schema = schemaInfo.name;
+
+  // Step 1: Grant firebasesuperuser access to the original owner
+  const uniqueTablesOwners = [...new Set(schemaInfo.tables.map((t) => t.owner))];
+  const grants = uniqueTablesOwners.map((owner) => `GRANT ${owner} TO ${FIREBASE_SUPER_USER}`);
+  await executeSqlCmdsAsSuperUser(options, instanceId, databaseId, grants, silent);
+
+  // Step 2: Using firebasesuperuser, setup reader and writer permissions on existing tables and setup default permissions for future tables.
+  const iamUser = (await getIAMUser(options)).user;
+  const projectNumber = await needProjectNumber(options);
+  const { user: fdcP4SAUser } = toDatabaseUser(getDataConnectP4SA(projectNumber));
+
+  const brownfieldSetupCmds = [
+    // Create and setup writer role permissions.
+    ...writerRolePermissions(databaseId, FIREBASE_SUPER_USER, schema),
+
+    // Create and setup reader role permissions.
+    ...readerRolePermissions(databaseId, FIREBASE_SUPER_USER, schema),
+
+    // Grant firebasewriter role to the current IAM user.
+    `GRANT "${firebasewriter(databaseId, schema)}" TO "${iamUser}"`,
+    // Grant firebaswriter to the FDC P4SA user
+    `GRANT "${firebasewriter(databaseId, schema)}" TO "${fdcP4SAUser}"`,
+  ];
+
+  // Add default Permissions
+  uniqueTablesOwners.forEach((owner) => {
+    brownfieldSetupCmds.push(...defaultPermissions(databaseId, schema, owner));
+  });
+
+  await executeSqlCmdsAsSuperUser(options, instanceId, databaseId, brownfieldSetupCmds, silent);
+}
+
+// Sets up all FDC roles (owner, writer, and reader).
+// Granting roles to users is done by the caller.
+// Returns true if schema is setup as greenfield
+export async function setupSQLPermissions(
+  instanceId: string,
+  databaseId: string,
+  schemaInfo: SchemaMetaData,
+  options: Options,
+  silent: boolean = false,
+): Promise<boolean> {
+  const schema = schemaInfo.name;
+  // Step 0: Check current user can run setup and upsert IAM / P4SA users
+  logger.info(`Attempting to Setup SQL schema "${schema}".`);
+  const userIsCSQLAdmin = await iamUserIsCSQLAdmin(options);
+  if (!userIsCSQLAdmin) {
+    throw new FirebaseError(`Only users with 'roles/cloudsql.admin' can setup SQL schemas.`);
```

Comment:
```suggestion
    throw new FirebaseError(`Missing required IAM permission to setup SQL schemas. SQL schema setup requires 'roles/cloudsql.admin' or an equivalent role.`);
```

Just to clarify that custom roles (or owner/editor) also works